cpp/include/cudf/detail/null_mask.cuh

```diff
@@ -199,253 +214,249 @@ __global__ void subtract_set_bits_range_boundaries_kernel(bitmask_type const* bi
-      bitmask_type slack_mask = set_least_significant_bits(num_slack_bits);
+    // Compute delta due to the preceding bits in the first word in the range.
+    size_type const first_num_slack_bits = intra_word_index(first_bit_index);
+    if (first_num_slack_bits > 0) {
+      bitmask_type const word       = bitmask[word_index(first_bit_index)];
+      bitmask_type const slack_mask = set_least_significant_bits(first_num_slack_bits);
       delta -= __popc(word & slack_mask);
     }
 
-    // compute delta due to the following bits in the last word in the range
-
-    num_slack_bits = (last_bit_index % word_size_in_bits) == 0
-                       ? 0
-                       : word_size_in_bits - intra_word_index(last_bit_index);
-    if (num_slack_bits > 0) {
-      bitmask_type word       = bitmask[word_index(last_bit_index)];
-      bitmask_type slack_mask = set_most_significant_bits(num_slack_bits);
+    // Compute delta due to the following bits in the last word in the range.
+    size_type const last_num_slack_bits = (last_bit_index % word_size_in_bits) == 0
+                                            ? 0
+                                            : word_size_in_bits - intra_word_index(last_bit_index);
+    if (last_num_slack_bits > 0) {
+      bitmask_type const word       = bitmask[word_index(last_bit_index)];
+      bitmask_type const slack_mask = set_most_significant_bits(last_num_slack_bits);
       delta -= __popc(word & slack_mask);
     }
 
+    // Update the null count with the computed delta.
     size_type updated_null_count = *(null_counts + range_id) + delta;
     *(null_counts + range_id)    = updated_null_count;
-
     range_id += blockDim.x * gridDim.x;
   }
 }
 
-// convert [first_bit_index,last_bit_index) to
-// [first_word_index,last_word_index)
-struct to_word_index : public thrust::unary_function<size_type, size_type> {
-  const bool _inclusive                 = false;
-  size_type const* const _d_bit_indices = nullptr;
-
-  /**
-   * @brief Constructor of a functor that converts bit indices to bitmask word
-   * indices.
-   *
-   * @param[in] inclusive Flag that indicates whether bit indices are inclusive
-   * or exclusive.
-   * @param[in] d_bit_indices Pointer to an array of bit indices
-   */
-  __host__ to_word_index(bool inclusive, size_type const* d_bit_indices)
-    : _inclusive(inclusive), _d_bit_indices(d_bit_indices)
-  {
-  }
-
-  __device__ size_type operator()(const size_type& i) const
+/**
+ * @brief Functor that converts bit segment indices to word segment indices.
+ *
+ * Converts [first_bit_index, last_bit_index) to [first_word_index,
+ * last_word_index). The flag `end_of_segment` indicates whether the bit is at
+ * the end of a segment, in which case the word index should be incremented for
+ * bits at the start of a word.
+ */
+struct bit_to_word_index {
+  bit_to_word_index(bool end_of_segment) : end_of_segment(end_of_segment) {}
```

Comment:
I think the inclusive/exclusive naming is clearer than `end_of_segment`.

Reply:
I'm afraid I too find `_inclusive` a little easier to grok. I would have thought `end_of_segment` would refer to both conditions together, i.e. `_inclusive || intra_word_index(i) == 0`.

Reply:
I wasn't convinced I made the right change here, so I appreciate the feedback. However, I think that `inclusive` is a slightly flawed name as well because we're really indicating whether we want the behavior of an inclusive _lower bound_ or an exclusive _upper bound_ (following the convention of defining ranges by `[start, end)`). I hoped that conveying the "lower/upper bound" concept in the name was clearer than the "inclusive/exclusive", since upper bounds are always exclusive by convention in C/C++ array programming. I'll revert this change to the name, but feel free to offer any other ideas.